test/extensions/filters/http/custom_response/custom_response_integration_test.cc

```diff
@@ -0,0 +1,353 @@
+
+  default_request_headers_.setHost("some.route");
+  auto response =
+      sendRequestAndWaitForResponse(default_request_headers_, 0, gateway_error_response_, 0);
+  // Verify we get the original error status code.
+  EXPECT_EQ("502", response->headers().getStatusValue());
+
+  EXPECT_EQ(1, test_server_->counter("http.config_test.downstream_rq_5xx")->value());
+}
+
+// Verify that we do not provide a custom response for a response that has
+// already been redirected by the custom response filter.
+// Verify that the route specific filter is picked if specified.
+TEST_P(CustomResponseIntegrationTest, NoRecursion) {
+
+  // Make the remote response for gateway_error policy return 401
+  auto fo1 = config_helper_.createVirtualHost("fo1.example");
+  fo1.mutable_routes(0)->set_name("fo1");
+  fo1.mutable_routes(0)->mutable_direct_response()->set_status(401);
+  fo1.mutable_routes(0)->mutable_direct_response()->mutable_body()->set_inline_string("fo1");
+  fo1.mutable_routes(0)->mutable_match()->set_prefix("/");
+  config_helper_.addVirtualHost(fo1);
+  custom_response_filter_config_.replace(custom_response_filter_config_.find("foo."), 4, "fo1.");
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Verify that a 401 response will trigger 400_response policy
+  default_request_headers_.setHost("some.route");
+  auto response =
+      sendRequestAndWaitForResponse(default_request_headers_, 0, unauthorized_response_, 0);
+  // Verify that we get the modified status value.
+  EXPECT_EQ("499", response->headers().getStatusValue());
+  EXPECT_EQ("not allowed", response->body());
+  EXPECT_EQ("x-bar",
+            response->headers().get(Http::LowerCaseString("foo"))[0]->value().getStringView());
+
+  // Verify that 400_response policy cannot be triggered by the
+  // gateway_error_response policy
+  default_request_headers_.setHost("some.route");
+  response = sendRequestAndWaitForResponse(default_request_headers_, 0, gateway_error_response_, 0);
+  // Verify we get status code for fo1.example
+  EXPECT_EQ("401", response->headers().getStatusValue());
+}
+
+// Verify that we can NOT intercept local replies sent during decode
+TEST_P(CustomResponseIntegrationTest, DecodeLocalReplyBeforeCER) {
+
+  // Add filter that sends local reply after.
+  filters_before_cer_.emplace_back(R"EOF(
+name: local-reply-during-decode
+)EOF");
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  default_request_headers_.setHost("some.route");
+  auto response = codec_client_->makeRequestWithBody(default_request_headers_, 10);
+  ASSERT_TRUE(response->waitForEndStream());
+  EXPECT_TRUE(response->complete());
+  // Verify we don't get a modified status value.
+  EXPECT_EQ("500", response->headers().getStatusValue());
+}
+
+// Verify that we can NOT intercept local replies sent during encode
+TEST_P(CustomResponseIntegrationTest, EncodeLocalReplyBeforeCER) {
```

Comment:
Is it true in all cases? Can local reply on the encode path be intercepted if it happens before the CER filter?

Reply:
If the filter that sent the local reply sends the same local reply again on redirection by the `RedirectPolicy` of the CustomResponse filter, the downstream will see the local reply, and not the response from the upstream after redirection.

Reply:
Actually, currently we cannot call `recreateStream` from within `sendLocalReply`. This could be fixed by leveraging the `onLocalReply` callback to return a new status which will cause the sendLocalReply to return early. I will send a PR for review following this one.